### PR TITLE
Layer mgr

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: servicemanager/servicemanager.go [hunk 1]
```diff
@@ -209,7 +211,12 @@ func (sm *ServiceManager) doInstallService(newService ServiceInfo, imageURL stri
 		}
 	}
 
-	if err = sm.prepareServiceFS(newService.ImagePath, newService.GID); err != nil {
+	footFSDigest, err := sm.prepareServiceFS(newService.ImagePath, newService.GID)
```
Comment (al1img) [inline servicemanager/servicemanager.go:228]: rootFSDigest

File: servicemanager/servicemanager.go [hunk 2]
```diff
@@ -209,7 +211,12 @@ func (sm *ServiceManager) doInstallService(newService ServiceInfo, imageURL stri
...
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = updateRootFSDigestInManifest(newService.ImagePath, footFSDigest); err != nil {
```
Comment (al1img) [inline servicemanager/servicemanager.go:233]: Could be part of prepareServiceFS?

File: servicemanager/image.go [hunk 3]
```diff
@@ -87,10 +87,27 @@ func validateUnpackedImage(installDir string) (err error) {
 	}
 
 	// validate service rootfs layer
-	if err = validateDigest(installDir, manifest.Layers[0].Digest); err != nil {
+	rootfsPath := path.Join(
+		installDir, "blobs", string(manifest.Layers[0].Digest.Algorithm()), manifest.Layers[0].Digest.Hex())
```
Comment (al1img) [inline servicemanager/image.go:91]: const for "blobs"?

File: servicemanager/servicemanager.go [hunk 4]
```diff
@@ -161,6 +162,19 @@ func (sm *ServiceManager) RevertService(service ServiceInfo) (retErr error) {
 	)
 }
 
+func (sm *ServiceManager) ValidateService(service ServiceInfo) (retErr error) {
```
Comment (al1img) [inline servicemanager/servicemanager.go:165]: remove retErr

File: utils/imageutils/imageutils.go [hunk 5]
```diff
@@ -30,14 +30,14 @@ import (
  * Types
  ******************************************************************************/
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply utils/imageutils/imageutils.go:34]: done

File: utils/imageutils/imageutils.go [hunk 6]
```diff
@@ -56,6 +59,37 @@ func CopyFile(source, destination string) (err error) {
 	return aoserrors.Wrap(err)
 }
 
+// ExtractPackageByUrl extract package by url and validate checksum.
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply utils/imageutils/imageutils.go:62]: done

File: layermanager/layermanager.go [hunk 7]
```diff
@@ -119,6 +118,90 @@ func (layermanager *LayerManager) GetLayersInfo() (info []LayerInfo, err error)
...
+			if err == nil {
+				err = aoserrors.Wrap(curErr)
+			}
+		}
+	}
+
+	return aoserrors.Wrap(err)
+}
+
+// GetLayerPathByDigest provied installed layer path by digest.
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply layermanager/layermanager.go:188]: done

File: layermanager/layermanager.go [hunk 8]
```diff
@@ -119,6 +118,90 @@ func (layermanager *LayerManager) GetLayersInfo() (info []LayerInfo, err error)
...
+func (layermanager *LayerManager) GetLayerPathByDigest(layerDigest string) (layerPath string, err error) {
+	layerPath, err = layermanager.layerInfoProvider.GetLayerPathByDigest(layerDigest)
+	if err != nil {
+		return "", aoserrors.Wrap(err)
+	}
+
+	return layerPath, nil
+}
+
+// GetLayerInfoByDigest get layers information by layer digest.
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply layermanager/layermanager.go:198]: done

File: layermanager/layermanager.go [hunk 9]
```diff
@@ -119,6 +118,90 @@ func (layermanager *LayerManager) GetLayersInfo() (info []LayerInfo, err error)
...
+				err = aoserrors.Wrap(curErr)
+			}
+		}
+	}
+
+	return aoserrors.Wrap(err)
+}
+
+// GetLayerPathByDigest provied installed layer path by digest.
+func (layermanager *LayerManager) GetLayerPathByDigest(layerDigest string) (layerPath string, err error) {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply layermanager/layermanager.go:189]: removed

File: database/database.go [hunk 10]
```diff
@@ -33,6 +33,7 @@ import (
 	log "github.com/sirupsen/logrus"
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply database/database.go:33]: fixed

File: servicemanager/image.go [hunk 11]
```diff
@@ -188,3 +217,36 @@ func getManifestChecksum(installDir string) (digest []byte, err error) {
...
+		r, err := open(file)
+		if err != nil {
+			return "", err
+		}
+
+		hf := sha256.New()
+
+		_, err = io.Copy(hf, r)
+
+		r.Close()
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply servicemanager/image.go:259]: impassable to implement your approach. In your case, we will have a lot of opened file descriptors

File: servicemanager/image.go [hunk 12]
```diff
@@ -188,3 +217,36 @@ func getManifestChecksum(installDir string) (digest []byte, err error) {
...
+func dirDigest(files []string, open func(string) (io.ReadCloser, error)) (string, error) {
+	h := sha256.New()
+
+	files = append([]string(nil), files...)
+
+	sort.Strings(files)
+
+	for _, file := range files {
+		if strings.Contains(file, "\n") {
+			return "", aoserrors.New("dirDigest: filenames with newlines are not supported")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply servicemanager/image.go:247]: done

File: servicemanager/servicemanager.go (hunk 2 above)
Comment (Eabramov84) [reply servicemanager/servicemanager.go:233]: from my point of view, manifest update should be saparate call

File: servicemanager/image.go (hunk 3 above)
Comment (Eabramov84) [reply servicemanager/image.go:91]: done

File: servicemanager/servicemanager.go (hunk 4 above)
Comment (Eabramov84) [reply servicemanager/servicemanager.go:165]: removed

File: servicemanager/servicemanager.go (hunk 1 above)
Comment (Eabramov84) [reply servicemanager/servicemanager.go:228]: done

File: layermanager/layermanager.go [hunk 13]
```diff
@@ -64,11 +63,22 @@ type LayerManager struct {
 
 // LayerInfoProvider provides API to add, remove or access layer information.
 type LayerInfoProvider interface {
-	AddLayer(digest, layerID, path, osVersion, vendorVersion, description string, aosVersion uint64) (err error)
+	AddLayer(layerInfo LayerInfo) (err error)
 	DeleteLayerByDigest(digest string) (err error)
 	GetLayerPathByDigest(digest string) (path string, err error)
-	GetLayersInfo() (layersList []*pb.LayerStatus, err error)
-	GetLayerInfoByDigest(digest string) (layer pb.LayerStatus, err error)
+	GetLayersInfo() (layersList []LayerInfo, err error)
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply layermanager/layermanager.go:75]: removed

**Review by al1img — APPROVED** [review]

**Pull request merged**